[RNMobile] Upgrade React Native `0.71.11`

fluiddot opened the pull request:
**⚠️ Please do not merge this PR until we ensure that the Gutenberg Mobile PRs referenced below are ready.**

## Related PRs
### ⚙️ Core
* [Gutenberg Mobile](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5874)
* [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android/pull/18613)
* [WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/20956)

<details><summary><h3>🤖 Android</h3></summary>

- [Gutenberg](https://github.com/WordPress/gutenberg/pull/51289)
- [Gutenberg Mobile](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5872)
- Update packages managed in `react-native-libraries-publisher`:
    - https://github.com/wordpress-mobile/react-native-libraries-publisher/pull/23
    - https://github.com/wordpress-mobile/react-native-libraries-publisher/pull/24
    - https://github.com/wordpress-mobile/react-native-libraries-publisher/pull/25
    - https://github.com/wordpress-mobile/react-native-libraries-publisher/pull/30
- Update forked packages
    - https://github.com/wordpress-mobile/react-native-hsv-color-picker/pull/10

</details><details><summary><h3>🍎 iOS</h3></summary>

- [Gutenberg](https://github.com/WordPress/gutenberg/pull/51386)
- [Gutenberg Mobile](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5881)
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/5924
- https://github.com/wordpress-mobile/react-native-reanimated/pull/26

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

</details>

## What?
<!-- In a few words, what is the PR actually doing? -->

Upgrades React Native to version ~~`0.71.8`~~`0.71.11`.

**UPDATE:** There were no breaking/major changes between `.8` and `.11` ([reference](https://react-native-community.github.io/upgrade-helper/?from=0.71.8&to=0.71.11)).

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This is part of a periodic effort to keep React Native up-to-date in the spirit of having the latest fixes and improvements.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Most of the changes have been applied following [the React Native upgrade helper](https://react-native-community.github.io/upgrade-helper/?from=0.69.4&to=0.71.11). Here is a list of the different commits applied and some notes:

* Update the different dependencies following the upgrade helper.
  * [Upgrade `react-native` dependency to version `0.71.8`](https://github.com/WordPress/gutenberg/pull/51303/commits/c70eb5b2c0eacde9e204180c0c81439989c72afa)
  * [Upgrade `react-native` dependency to version `0.71.11`](https://github.com/WordPress/gutenberg/pull/51303/commits/e0cab5942ee5760216117537a7bb0edb95b77d7a)
  * [Upgrade `@babel/runtime` dependency](https://github.com/WordPress/gutenberg/pull/51303/commits/13ad6087d5037fb40f804820d60e2a13ae107cb1) and 
* [Upgrade `cocoapods` gem](https://github.com/WordPress/gutenberg/pull/51303/commits/256be9c268c0ad8f4d5792046bbd2159441a4236)
* [Re-apply `react-devtools-core` patch to new version](https://github.com/WordPress/gutenberg/pull/51303/commits/213073d7a3d385d08c9f38cbac742dbb4fdfe28b)
* Update third-party packages.
  * [Update `@react-navigation/native` package to version 6.0.14](https://github.com/WordPress/gutenberg/pull/51303/commits/5b485736e3e482b13f9d74ee32ef7c8e649dec40)
  * [Update `react-native-reanimated` to version 2.17.0](https://github.com/WordPress/gutenberg/pull/51303/commits/ba8496a30cfc1a3648480dbdd4fc0cc2c6d235e3)
  * [Update `react-native-gesture-handler` to version 2.10.2](https://github.com/WordPress/gutenberg/pull/51303/commits/ca395c2515bc22fa0cd244992970fda08cbed313)
  * [Upgrade `react-native-linear-gradient` to version 2.7.3](https://github.com/WordPress/gutenberg/pull/51303/commits/556b3efe30cd29db8de0df5cc9fd68903775b525)
* Apply changes to fix tests that began to fail after the upgrade.
  * [Update jest snapshots with new a11y values](https://github.com/WordPress/gutenberg/pull/51303/commits/cbdb7ff7cf7968fc528318838f83a6ccca683b25)
  * ~~[Mock `Linking.addEventListener` function](https://github.com/WordPress/gutenberg/pull/51303/commits/bd0750b43387509c525d70e002f2fd5441025eae)~~ Replaced by [Mock return value of Linking addEventListener](https://github.com/WordPress/gutenberg/pull/51303/commits/97265c70ebe022868835bf476928e069e131461d)
  * [Update tests that fail due to the use of debounce and link suggestions](https://github.com/WordPress/gutenberg/pull/51303/commits/6cac580e44c4cc09c91994aeeb24f7daed6f4328)
  * [Fix `MediaUpload` component test](https://github.com/WordPress/gutenberg/pull/51303/commits/311ed3ffe1b28c19feeb3a66202a05c9af027b3b)
  * [Fix `act` warnings produced during block insertion](https://github.com/WordPress/gutenberg/pull/51303/commits/2c0e048f2ddd8bd8af9410469c073c3597fc3dc8)
  * [Fix `act` warnings in Columns block tests](https://github.com/WordPress/gutenberg/pull/51303/commits/809b82dd6cf6f2070f11468b8861cba292bc924d)
  * [Fix `act` warnings in List block tests](https://github.com/WordPress/gutenberg/pull/51303/commits/3f7357400f8b6191345b3755cae04ae742ad73d1)
* [iOS changes](https://github.com/WordPress/gutenberg/pull/51303/commits/305db7fc7991460ae84d11ca78962df1b941a3e5)
* [Android changes](https://github.com/WordPress/gutenberg/pull/51303/commits/39570ce964289187802343ed0bac582c86cbf5ff)
* E2E tests fixes
    * [Avoid exception in E2E tests when typing an empty string on Android](https://github.com/WordPress/gutenberg/pull/51303/commits/ad188ac4f2b05bc72f478c85e66bba795deba82f)
    * [Update a11y id queries for Android E2E tests](https://github.com/WordPress/gutenberg/pull/51303/commits/9b8d7be878c557e7dee57ce81701c88da62cc64c)
    * [Update button inline appender query for Android E2E tests](https://github.com/WordPress/gutenberg/pull/51303/commits/de279689b862b383af1eaf36e9282327ae31ec80)
    * [Unify press keycode function for E2E tests](https://github.com/WordPress/gutenberg/pull/51303/commits/b6f5fba4ecce61dc4d079f00a3371319ae55c5e6)
* Fixes to critical issues spotted during testing
    * [fix: Cover focal point drag handle visibility](https://github.com/WordPress/gutenberg/pull/51303/commits/470af373ffa5f83cd581f56c699de3d1b15d6e6b)
    * [Use Reanimated in bottom sheet height animation](https://github.com/WordPress/gutenberg/pull/51303/commits/0411df13e1e8b832d8011e4fafe34436cd8fc773)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
* Smoke test the editor.
* Test areas related to the third-party packages updated.
  * Reanimated and Gesture handler:
    * Observe that dragging and dropping blocks work as expected.
  * React Navigation:
    * Observe that navigation in screens within the bottom sheet work as expected (e.g. Color settings).
    * Observe that navigation in the Help & Support section works as expected.
  * Linear gradient
    * Observe that gradient colors are rendered as expected, both linear and radial modes. This can be tested using the Cover block.
    * Since this library is also used in the Color picker component, observe that picking a custom color also works as expected. To test this in the demo version, we might need to apply the following patch.
    ```
    diff --git forkSrcPrefix/packages/components/src/mobile/color-settings/palette.screen.native.js forkDstPrefix/packages/components/    src/mobile/color-settings/palette.screen.native.js
    index bc7187fd092b8ca5a69c81f7cdb2bd1861c7d6e5..06fefdc8c2e07aeb6016297e89910d32c74b9324 100644
    --- forkSrcPrefix/packages/components/src/mobile/color-settings/palette.screen.native.js
    +++ forkDstPrefix/packages/components/src/mobile/color-settings/palette.screen.native.js
    @@ -191,7 +191,7 @@ const PaletteScreen = () => {
     
     					return (
     						<ColorPalette
    -							enableCustomColor={ enableCustomColor }
    +							enableCustomColor={ true }
     							label={ palette.name }
     							key={ paletteKey }
     							setColor={ setColor }
    ```
* Check all CI jobs pass.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->
N/A

fluiddot committed:
Upgrade `react-native` dependency

fluiddot committed:
Upgrade `@babel/runtime` dependency

fluiddot committed:
Upgrade `metro-react-native-babel` preset and transformer dependencies

fluiddot committed:
Upgrade `cocoapods` gem

fluiddot committed:
Re-apply `react-devtools-core` patch to new version

fluiddot committed:
Update jest snapshots with new a11y values

fluiddot committed:
Mock `Linking.addEventListener` function

`Linking.removeEventListener` has been removed in RN `0.71`. The library is mocked by default but doesn't return the `remove` function when calling `addEventListener`.

fluiddot committed:
Update tests that fail due to use of debounce and link suggestions

fluiddot committed:
Fix `MediaUpload` component test

fluiddot committed:
Update `@react-navigation/native` package to version `6.0.14`

fluiddot committed:
Update `react-native-reanimated` to version `2.17.0`

fluiddot committed:
Update `react-native-gesture-handler` to version `2.10.2`

fluiddot commented on packages/block-editor/src/components/block-draggable/test/index.native.js:
```diff
@@ -437,7 +437,7 @@ describe( 'BlockDraggable', () => {
 			// displays a warning message.
 			expect( console ).toHaveWarnedWith(
-				'[react-native-gesture-handler] You have to use react-native-reanimated in order to control the state of the gesture.'
+				'[Reanimated] You can not use setGestureState in non-worklet function.'
```
The previous warning is no longer shown when running the tests, seems that now a different one is displayed instead.

fluiddot commented on packages/block-editor/src/components/block-mover/test/__snapshots__/index.native.js.snap:
```diff
@@ -51,7 +51,19 @@ exports[`Block Mover Picker should match snapshot 1`] = `
       accessibilityRole="button"
       accessibilityState={
         {
+          "busy": undefined,
+          "checked": undefined,
           "disabled": false,
+          "expanded": undefined,
+          "selected": undefined,
+        }
+      }
+      accessibilityValue={
+        {
+          "max": undefined,
+          "min": undefined,
+          "now": undefined,
+          "text": undefined,
```
These new a11y values have been added in RN 0.71 ([reference](https://github.com/facebook/react-native/commit/98d84e571df64b00e1ed3484923a1d8169dcfda1)).

fluiddot commented on packages/block-editor/src/components/media-upload/test/index.native.js:
```diff
@@ -44,6 +44,7 @@ describe( 'MediaUpload component', () => {
 			const wrapper = render(
 				<MediaUpload
 					allowedTypes={ [ mediaType ] }
+					onSelectURL={ jest.fn() }
```
This fixes a failure I encountered when running the tests. The error was related to missing the option "Insert from URL" on the Image and Audio block.

After reviewing the logic of `MediaUpload` component, seems the error is legit, as the URL insertion option is only available when passing the `onSelectURL` prop. Now I'm curious about how these tests passed before 🤔 .

https://github.com/WordPress/gutenberg/blob/b52d61ed6c1307e9ccf0eedaddafe52c06033dbd/packages/block-editor/src/components/media-upload/index.native.js#L138C24-L145

fluiddot commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -26,6 +28,19 @@ import { ENTER } from '@wordpress/keycodes';
  */
 import Paragraph from '../edit';
 
+// Mock debounce to prevent potentially belated state updates.
+jest.mock( '@wordpress/compose/src/utils/debounce', () => ( {
+	debounce: ( fn ) => {
+		fn.cancel = jest.fn();
+		return fn;
+	},
+} ) );
```
The link suggestions are fetched using `debounce`. Since we are not testing the suggestions in this test suites, I disabled debounce to avoid belated state updates.

https://github.com/WordPress/gutenberg/blob/b52d61ed6c1307e9ccf0eedaddafe52c06033dbd/packages/components/src/mobile/link-picker/link-picker-results.native.js#L75

fluiddot commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -26,6 +28,19 @@ import { ENTER } from '@wordpress/keycodes';
+		return fn;
+	},
+} ) );
+// Mock link suggestions that are fetched by the link picker
+// when typing a search query.
+jest.mock( '@wordpress/core-data/src/fetch', () => ( {
+	__experimentalFetchLinkSuggestions: jest.fn().mockResolvedValue( [ {} ] ),
+} ) );
```
I noticed that the Link picker component is making requests to fetch the link suggestions. This was causing a failure in the tests because, although `apiFetch` is mocked, it's not returning a promise. This logic is not being tested in these test suites, so I simply mock it.

https://github.com/WordPress/gutenberg/blob/b52d61ed6c1307e9ccf0eedaddafe52c06033dbd/packages/components/src/mobile/link-picker/link-picker-results.native.js#L42-L45

https://github.com/WordPress/gutenberg/blob/b52d61ed6c1307e9ccf0eedaddafe52c06033dbd/packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js#L101-L119

https://github.com/WordPress/gutenberg/blob/b52d61ed6c1307e9ccf0eedaddafe52c06033dbd/test/native/setup.js#L73-L78

fluiddot commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -238,35 +253,37 @@ describe( 'Paragraph block', () => {
 			screen.getByPlaceholderText( 'Add link text' ),
 			'WordPress'
 		);
-		jest.useFakeTimers();
-		fireEvent.press( screen.getByLabelText( 'Apply' ) );
-		// Await link picker navigation delay
-		act( () => jest.runOnlyPendingTimers() );
```
I updated this part to use the helper `withFakeTimers` which already sets fake timers, as well as other tweaks.

fluiddot commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -238,35 +253,37 @@ describe( 'Paragraph block', () => {
+				screen.getByPlaceholderText( 'Search or type URL' ),
+				'wordpress.org'
+			)
+		);
```
This `await` solves an `act` warning caused by link suggestions.

https://github.com/WordPress/gutenberg/blob/b52d61ed6c1307e9ccf0eedaddafe52c06033dbd/packages/components/src/mobile/link-picker/link-picker-results.native.js#L57-L68

fluiddot commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -238,35 +253,37 @@ describe( 'Paragraph block', () => {
+		fireEvent.press( screen.getByLabelText( 'Link' ) );
+
 		fireEvent.changeText(
 			screen.getByPlaceholderText( 'Add link text' ),
```
The link text is set before adding the link, otherwise, it's populated with the link URL.

dcalhoun replied:
Was populating the text via the link URL causing issue or done merely for clear presentation?

fluiddot replied:
The original approach was updating the Link text field while the Link picker modal was open. This is not possible when using the editor, but could be done in the test as the element is accessible. 

https://github.com/WordPress/gutenberg/blob/e2cfaaaa7a4899d631ad67c09b283b173a093a35/packages/block-library/src/paragraph/test/edit.native.js#L242-L256

| Link settings | Link picker modal |
|--------|--------|
| <img src=https://github.com/WordPress/gutenberg/assets/14905380/5b97e812-33b7-4654-8f2a-c3b89fe7b7ae> | <img src=https://github.com/WordPress/gutenberg/assets/14905380/2a1f1d46-e6e6-4525-baa3-06a7bb7c5781> | 

That's the main reason why I changed the order of the actions.

fluiddot commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -287,31 +304,32 @@ describe( 'Paragraph block', () => {
+		fireEvent.press(
+			screen.getByLabelText( 'Link to, Search or type URL' )
+		);
+		await waitForModalVisible(
+			screen.getByTestId( 'link-settings-modal' )
+		);
```
This wait will ensure that the modal is visible when typing an URL.

dcalhoun replied:
Would awaiting the presence of the input itself not accomplish the same outcome with a more pointed query?

fluiddot replied:
Yeah, that would be another option, we could do the following instead:
```
const typeURLInput = await waitFor( () =>
	screen.getByPlaceholderText( 'Search or type URL' )
);
// Wait for side effects produced by link suggestions
await act( () =>
	fireEvent.changeText( typeURLInput, 'wordpress.org' )
);
```

I'll proceed to update this with your suggestion.

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/9f2edf4c23a086cedfc5acfdccbd9e820495ce68.

fluiddot commented on packages/components/src/mobile/link-settings/test/edit.native.js:
```diff
@@ -34,9 +47,7 @@ describe.each( [
-				<div class="wp-block-button">
-					<a class="wp-block-button__link" style="border-radius:5px">Link</a>
-				</div>
+				<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" style="border-radius:5px">Link</a></div>
```
This HTML has been updated as its format was outdated.

dcalhoun replied:
I suppose this is a good example of the benefit of leveraging UI instead of initial HTML. I wonder if there is are benefits to relying upon initial HTML. Test speed possibly? I'm unsure of if the impact is negligible.

fluiddot replied:
I think the cases where we use initial HTML are mainly driven to start the editor in a specific state, similar to opening a post with content. This has the downside of using fixed versions of the block's HTML code, which as in this case, it would eventually lead to being outdated.

It's likely that adding the block and setting the content is slower than using the initial HTML. I haven't reviewed thoroughly these test cases to determine if could use a different approach, my main goal in this PR was to simply fix the issues to unblock the RN upgrade. Nevertheless, we could improve this in another PR if we feel the need to.

dcalhoun replied:
> I haven't reviewed thoroughly these test cases to determine if could use a different approach, my main goal in this PR was to simply fix the issues to unblock the RN upgrade. Nevertheless, we could improve this in another PR if we feel the need to.

Definitely. No intention from my side to modify the testing approach in this PR. Merely shared the thought for general consideration.

fluiddot commented on packages/components/src/mobile/link-settings/test/edit.native.js:
I applied several updates here to solve different failures I encountered when running the tests, as well as `act` warnings. In general, I followed the approach of adding `await` statements on actions that have deferred state updates (like getting Clipboard data or fetching link suggestions).

I also migrated some of the `find` queries to `get`, for the sake of aligning with other integration tests.

dcalhoun replied:
> I also migrated some of the `find` queries to `get`, for the sake of aligning with other integration tests.

This a bit confusing to me. While I recognize the benefit of consistency of approach, it appears this (A) pointedly reverts the inverse change in [`705ab4`](https://github.com/WordPress/gutenberg/commit/705ab44f2c2797d75740533a85030c194b4fb040#diff-f31c53b09b870a8c2290951101c970a557c9ad2ab8dbd65d97e75b2b1b356831) and (B) is likely counter to the effort of removing our custom `waitFor` in https://github.com/WordPress/gutenberg/pull/46735. 

Do you believe we should avoid migrating from `find` to `get` if our overarching goal is to remove our custom `waitFor`? My perception is that `find` is more convenient and the recommend approach for async queries.  

---

This thought applies to other test files in this PR that appear to migrate from `find` to `get` queries as well.

fluiddot replied:
> This a bit confusing to me. While I recognize the benefit of consistency of approach, it appears this (A) pointedly reverts the inverse change in [`705ab4`](https://github.com/WordPress/gutenberg/commit/705ab44f2c2797d75740533a85030c194b4fb040#diff-f31c53b09b870a8c2290951101c970a557c9ad2ab8dbd65d97e75b2b1b356831) and (B) is likely counter to the effort of removing our custom `waitFor` in #46735.

Oh, good catch, I overlooked that this was updated when upgrading React 18 🤦. I tried to follow [the documentation of integration tests](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/react-native/integration-test-guide.md#use-of-waitfor), specifically:
> NOTE: The react-native-testing-library package provides the query* and find* functions for this purpose too, but we should avoid using them for now because there’s a known issue that would make the test fail.

In any case, I agree that with upcoming changes in https://github.com/WordPress/gutenberg/pull/46735 and the fact that this is reverting recent changes, we shouldn't apply these changes. I'll revert all changes related to using `waitFor` I made related to this 👍 .

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/a9d4165b106d925a49bc08fe0afc471e7c2d6070. Note that I only reverted the usages of `waitFor`, I've left some of the changes from `find` to `get`, as we can query them synchronously.

fluiddot commented on patches/@react-navigation+native+5.7.0.patch:
This patch is no longer necessary as `@react-navigation/native` has been updated to `6.x` ([reference](https://github.com/react-navigation/react-navigation/commit/08d6e14477acc757fcdf945350a789e5fbae822b)).

fluiddot commented on patches/react-devtools-core+4.27.8.patch:
This patch replaces the one made for the previous version of `react-devtools-core` (https://github.com/WordPress/gutenberg/pull/51303/files#diff-efdeee0f12bb2896ff685c25c7b899da1f11686b6d333de8c33265db624b870e).

fluiddot commented on test/native/setup.js:
```diff
@@ -242,6 +242,11 @@ jest.mock(
 jest.mock( 'react-native/Libraries/ActionSheetIOS/ActionSheetIOS', () => ( {
 	showActionSheetWithOptions: jest.fn(),
 } ) );
+jest.mock( 'react-native/Libraries/Linking/Linking', () => {
+	return {
+		addEventListener: jest.fn( () => ( { remove: jest.fn() } ) ),
+	};
+} );
```
`Linking.removeEventListener` has been removed in RN `0.71` ([reference](https://reactnative.dev/docs/linking)). The [library is mocked by default](https://github.com/facebook/react-native/blob/d2a873956520806e434084d2d48dc0a4ee913f07/jest/setup.js#L181) but doesn't return the `remove` function when calling `addEventListener`.

dcalhoun replied:
Thinking "out loud:" is there a way to avoid mocking and merely update the returned mock value? Not necessarily confident this is necessary or better, but I wondered if it is helpful to avoid re-mocking a module already managed by RN core. WDYT? 

```js
Linking.addEventListener.mockReturnValue({ remove: jest.fun() });
```

fluiddot replied:
Yep, definitely, we could only mock the return value of `addEventListener`.

fluiddot replied:
I applied your suggestion in https://github.com/WordPress/gutenberg/pull/51303/commits/97265c70ebe022868835bf476928e069e131461d.

fluiddot committed:
Fix `act` warnings produced during block insertion

fluiddot committed:
Fix `act` warnings in Columns block tests

fluiddot committed:
Fix `act` warnings in List block tests

fluiddot commented on test/native/integration-test-helpers/add-block.js:
```diff
@@ -44,7 +44,19 @@ export const addBlock = async (
 	} );
 
-	fireEvent.press( await waitFor( () => getByText( blockName ) ) );
+	const blockButton = await waitFor( () => getByText( blockName ) );
+	// Blocks can perform belated state updates after they are inserted.
+	// To avoid potential `act` warnings, we ensure that all timers and queued
+	// microtasks are executed.
+	await withFakeTimers( async () => {
+		fireEvent.press( blockButton );
+		// Run all timers, in case any performs a state updates.
+		// Column block example: https://t.ly/NjTs
+		act( () => jest.runOnlyPendingTimers() );
+		// Let potential queued microtasks (like Promises) to be executed.
+		// Inner blocks example: https://t.ly/b95nA
+		await act( async () => {} );
+	} );
```
I noticed that with RN `0.71.8`, there was a dramatic increase of `act` warnings when running tests ([reference](https://github.com/WordPress/gutenberg/actions/runs/5214263615/jobs/9410316091#step:5:491)). I investigated further and found out that most of them were related to block insertion, especially for blocks with inner blocks. I identified belated state updates produced by microtasks ([reference](https://github.com/WordPress/gutenberg/blob/05b94d1ff57027d53e4fed98a42cf80ca65f42aa/packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js#L132-L143)). With this approach, we'll ensure that all state updates potentially produced during block creation are handled 🤞 .

@dcalhoun as we worked closely on [a similar issue with `act` warnings](https://github.com/WordPress/gutenberg/pull/38052). I'd appreciate your input in this workaround. Thanks 🙇

dcalhoun commented on test/native/integration-test-helpers/add-block.js:
```diff
@@ -44,7 +44,19 @@ export const addBlock = async (
+	// microtasks are executed.
+	await withFakeTimers( async () => {
+		fireEvent.press( blockButton );
+		// Run all timers, in case any performs a state updates.
+		// Column block example: https://t.ly/NjTs
+		act( () => jest.runOnlyPendingTimers() );
```
I am hesitant to await this nuanced UI logic in a test helper. This feels like something that should be accounted for within the Columns test file. 

I am unaware of a specific example right now, but I could see abstracting the timers to this top-level helper confusing future test writers. E.g. if my subject block transitions from one state to another over a period of time, i.e. with `setTimeout`, this global helper would prevent me from asserting (or simply understanding and recognizing) that transition. 

Are there `act` examples outside of the Column block of which you are aware? Does it make sense to relocate this to the block tests that need it?

fluiddot replied:
I haven't checked if other blocks use timers to update their local state when created. I agree that, if this is only needed for a specific block, we could execute in the associated test files instead of in a generic way.

My purpose with this approach was to simplify the block insertion, as I feel that it won't be very common to assert potential state updates that happen during the insertion. I assumed that all of them acted as part of the `addBlock` helper. However, in case we need to, we could run the timers conditionally via a configuration parameter.

I'll check if this is needed in more blocks, and depending on that, we could decide if we rather run the timers generally here or in each test 👍 .

dcalhoun replied:
Your expressed purpose makes a lot of sense. I recognize that our testing environment needs to find a balance between convenience and comprehensibility. 

To be clear, I am not opposed to inserting this global timer into the `addBlock` helper. I merely wanted to note its potential for breeding confusion due to the obscure nature of this timers run. The inline comments will likely help to mitigate that, provided the test author reads the source of `addBlock` when using it. 

Please feel free to leave this implementation as-is. A configuration parameter could arguably be unnecessary complexity.

dcalhoun commented on test/native/integration-test-helpers/add-block.js:
```diff
@@ -44,7 +44,19 @@ export const addBlock = async (
+		// Run all timers, in case any performs a state updates.
+		// Column block example: https://t.ly/NjTs
+		act( () => jest.runOnlyPendingTimers() );
+		// Let potential queued microtasks (like Promises) to be executed.
+		// Inner blocks example: https://t.ly/b95nA
+		await act( async () => {} );
```
From reviewing the example inner block state changes, it makes sense to abstract this microtask flush, from my perspective. 

The changing state is already abstract, not incredibly block-specific, and does not always result in changes that are queryable in the UI. 

| Setting | Type | Queryable Result |
| - | - | - |
| `allowedBlocks` | `WPBlockType[]` | Seemingly unavailable. |
| `orientation` | `"vertical" \| "horizontal"` | Block mover arrows. |
| `prioritizedInserterBlocks` | `WPBlockType[]` | Seemingly unavailable |
| `templateLock` | `boolean` | Lock icon? |

dcalhoun commented on packages/components/src/mobile/link-settings/test/edit.native.js:
```diff
@@ -112,21 +123,25 @@ describe.each( [
 				if ( type === 'core/image' ) {
-					fireEvent.press(
-						await subject.findByLabelText( /Custom URL/ )
+					// Wait for side effects produced by Clipboard and link suggestions
+					await act( () =>
+						fireEvent.press(
+							subject.getByLabelText( /Custom URL/ )
+						)
 					);
```
My understanding is that `fireEvent` [includes](https://callstack.github.io/react-native-testing-library/docs/api#act) `act` itself. While I recognize there are scenarios we must wait additional updates ourself, I wonder if there is a more communicative way of doing so that does not wrap `fireEvent`. 

Is there a visible update for which we can query? What changes with the Clipboard promise resolves?

fluiddot replied:
> My understanding is that `fireEvent` [includes](https://callstack.github.io/react-native-testing-library/docs/api#act) `act` itself. While I recognize there are scenarios we must wait additional updates ourself, I wonder if there is a more communicative way of doing so that does not wrap `fireEvent`.

Yes, that's correct. The `await act` used here is to wait for belated state updates produced by getting clipboard data and link suggestions, which both happen asynchronously using promises.

Alternatively, we could do instead:
```
fireEvent.press( subject.getByLabelText( /Custom URL/ ) );
// Wait for side effects produced by Clipboard and link suggestions
await act( async () => {} );
```

dcalhoun replied:
Given your proposed alternative of a generic `await act( async () => {} )`, I presume there is no queryable/visible change we could await? E.g. `await findByText( /Idealistic element from Clipboard or link suggestion resolution/ )`.

fluiddot replied:
Ah, sorry, I overlooked the last paragraph 🙇 .

> Is there a visible update for which we can query? What changes with the Clipboard promise resolves?

I think there's none for this case, as the test is checking that there's no visual element displayed in the Link Picker:

https://github.com/WordPress/gutenberg/blob/3f7357400f8b6191345b3755cae04ae742ad73d1/packages/components/src/mobile/link-settings/test/edit.native.js#L149-L152

dcalhoun replied:
Ah, thank you for sharing. 

From reviewing the test and subject a bit further, it appears we may have an opportunity to make this `act` discussion moot and also avoid unnecessary re-renders. The fact that this component uses an object in a single `useState` makes it far more complicated. 

If we refactor to use separate `useState` for each value, then it would simplify the component logic and avoid unnecessary re-renders due to new object creation. It would look something like the following diff. **To be clear, I am not necessarily proposing we must do this within this PR.**

<details><summary>Diff: Split State; Avoid New Objects</summary>

```diff
diff --git a/packages/components/src/mobile/link-picker/index.native.js b/packages/components/src/mobile/link-picker/index.native.js
index ffde5b28c2..be7f70c195 100644
--- a/packages/components/src/mobile/link-picker/index.native.js
+++ b/packages/components/src/mobile/link-picker/index.native.js
@@ -57,10 +57,8 @@ export const LinkPicker = ( {
 	onLinkPicked,
 	onCancel: cancel,
 } ) => {
-	const [ { value, clipboardUrl }, setValue ] = useState( {
-		value: initialValue,
-		clipboardUrl: '',
-	} );
+	const [ value, setValue ] = useState( initialValue );
+	const [ clipboardUrl, setClipboardUrl ] = useState( '' );
 	const directEntry = createDirectEntry( value );
 
 	// The title of a direct entry is displayed as the raw input value, but if we
@@ -74,7 +72,8 @@ export const LinkPicker = ( {
 	};
 
 	const clear = () => {
-		setValue( { value: '', clipboardUrl } );
+		setValue( '' );
+		setClipboardUrl( '' );
 	};
 
 	const omniCellStyle = usePreferredColorSchemeStyle(
@@ -89,8 +88,10 @@ export const LinkPicker = ( {
 
 	useEffect( () => {
 		getURLFromClipboard()
-			.then( ( url ) => setValue( { value, clipboardUrl: url } ) )
-			.catch( () => setValue( { value, clipboardUrl: '' } ) );
+			.then( ( url ) => {
+				setClipboardUrl( url );
+			} )
+			.catch( () => setClipboardUrl( '' ) );
 		// Disable reason: deferring this refactor to the native team.
 		// see https://github.com/WordPress/gutenberg/pull/41166
 		// eslint-disable-next-line react-hooks/exhaustive-deps
@@ -116,7 +117,7 @@ export const LinkPicker = ( {
 					autoCorrect={ false }
 					keyboardType="url"
 					onChangeValue={ ( newValue ) => {
-						setValue( { value: newValue, clipboardUrl } );
+						setValue( newValue );
 					} }
 					onSubmit={ onSubmit }
 					/* eslint-disable-next-line jsx-a11y/no-autofocus */
diff --git a/packages/components/src/mobile/link-settings/test/edit.native.js b/packages/components/src/mobile/link-settings/test/edit.native.js
index 6ccb6ae630..9e7607f79b 100644
--- a/packages/components/src/mobile/link-settings/test/edit.native.js
+++ b/packages/components/src/mobile/link-settings/test/edit.native.js
@@ -125,15 +125,13 @@ describe.each( [
 				fireEvent.press( block );
 				fireEvent.press( subject.getByLabelText( 'Open Settings' ) );
 				// Wait for side effects produced by Clipboard
-				await act( () =>
-					fireEvent.press(
-						subject.getByLabelText(
-							`Link to, ${
-								type === 'core/image'
-									? 'None'
-									: 'Search or type URL'
-							}`
-						)
+				fireEvent.press(
+					subject.getByLabelText(
+						`Link to, ${
+							type === 'core/image'
+								? 'None'
+								: 'Search or type URL'
+						}`
 					)
 				);
 				if ( type === 'core/image' ) {

```

</details>

I am fine proceeding with the code as-is. It may be worth leaving an inline comment in the test capturing this potential refactor negating the `act` call.

fluiddot replied:
> From reviewing the test and subject a bit further, it appears we may have an opportunity to make this `act` discussion moot and also avoid unnecessary re-renders. The fact that this component uses an object in a single `useState` makes it far more complicated.

Nice observations. I agree, this part could be improved by avoiding a single `useState` for both values.

> If we refactor to use separate `useState` for each value, then it would simplify the component logic and avoid unnecessary re-renders due to new object creation. It would look something like the following diff. **To be clear, I am not necessarily proposing we must do this within this PR.**
>
> I am fine proceeding with the code as-is. It may be worth leaving an inline comment in the test capturing this potential refactor negating the `act` call.

Completely agree. Adding a comment explaining the potential refactor would be helpful, we could also create a new GitHub issue with all this information. WDYT?

dcalhoun replied:
A new issue makes sense. I am happy to manage this, potentially even "short circuit" the process and open PR with the changes directly. Consider it on my to-do list.

dcalhoun replied:
I opened https://github.com/WordPress/gutenberg/pull/51613 to resolve this specific issue, but plan to merge it _after_ this React Native upgrade PR is merged. Doing so will simplify testing the new PR due to upgrades to the testing packages that uncovered this the `act` warnings. 

Relatedly, I added 70dbcc60a1fdf4eb1f7ccb9eedf05e037b141bc0 to this React Native upgrade branch, which addresses a similar `act` warning, but did not require refactoring the `LinkPicker` component itself. So, I figure it is safe to include in this PR.

dcalhoun commented on packages/block-library/src/columns/test/edit.native.js:
```diff
@@ -293,6 +294,10 @@ describe( 'Columns block', () => {
 		// Get the first column
 		const firstColumnBlock = await getBlock( screen, 'Column' );
 		fireEvent.press( firstColumnBlock );
+		// Inner blocks batch store updates with microtasks.
+		// To avoid `act` warnings, we let queued microtasks to be executed.
+		// Reference: https://t.ly/b95nA
+		await act( async () => {} );
```
I suppose this could eventually be replaced with `addBlock` if the helper retains its version of this.

fluiddot replied:
Yes, although this test case doesn't use `addBlock` but initial HTML ([reference](https://github.com/WordPress/gutenberg/pull/51303/files#diff-808bc213c23a4817b913208333ab10e7410cc791fe1fd73f89d4b1e742c10099R275-R277)).

dcalhoun replied:
Ah, I misinterpreted the diff summary as being related to _adding_ an inner block. The reality is that this line relates to modifying an inner block already existing in the initial HTML. Apologies of the noise.

dcalhoun commented on packages/block-library/src/list/test/edit.native.js:
```diff
@@ -476,6 +477,10 @@ describe( 'List block', () => {
 			preventDefault() {},
 			keyCode: BACKSPACE,
 		} );
+		// Inner blocks batch store updates with microtasks.
+		// To avoid `act` warnings, we let queued microtasks to be executed.
+		// Reference: https://t.ly/b95nA
+		await act( async () => {} );
```
I suppose this could eventually be replaced with `addBlock` if the helper retains its version of this.

fluiddot replied:
Yes, although this test case doesn't use `addBlock` but initial HTML ([reference](https://github.com/WordPress/gutenberg/pull/51303/files#diff-52bc11cb1e553aa6c760699ba925f3627336ce8d841939d4a9992726f60887dbR439-R457)).

dcalhoun replied:
Ah, I misinterpreted the diff summary as being related to _adding_ an inner block. The reality is that this line relates to modifying an inner block already existing in the initial HTML. Apologies of the noise.

fluiddot committed:
Upgrade `react-native` dependency to version `0.71.11`

It also upgrades `metro-react-native-babel` dependencies following the upgrade helper.

fluiddot committed:
Mock return value of Linking `addEventListener`

We only need to mock the return the value, hence we don't need to mock the entire library.

fluiddot commented:
Heads up that I upgraded React Native to a new patch version (`0.71.11`) in order to incorporate a fix to avoid a crash on Android ([reference](https://github.com/facebook/react-native/commit/f7e35d4ef7d68d06fba1439c0aa6d9ed05b58a7f)). I'll update the PR's title and description.

fluiddot committed:
Remove `waitForModalVisible` usage in Paragraph block tests

fluiddot committed:
Remove `waitFor` usage in Link settings tests

dcalhoun committed:
test: Fix act warning by awaiting LinkPicker loading indicator removal

The loading indicator is displayed and subsequently removed once the
suggestion fetches resolve. Explicitly awaiting this element's removal
fixes the `act` warnings.

dcalhoun committed:
build: Update react-native-safe-area-context to 4.6.3

dcalhoun committed:
build: Upgrade react-native-screens to 3.22.0

dcalhoun committed:
build: Upgrade react-native-svg to 13.9.0

Based on the release notes breaking changes, we should look out for odd
sizing or display of icons, particularly on Android.

dcalhoun committed:
build: Upgrade @react-native-masked-view/masked-view to 0.2.9

dcalhoun committed:
build: Upgrade @react-native-clipboard/clipboard to 1.11.2

dcalhoun committed:
build: Upgrade react-native-modal to 13.0.1

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into rnmobile/upgrade/react-native-0.71.8

dcalhoun committed:
test: Update link modal snapshot

This change is a result of applying new props from the RN upgrade to a
newly introduced snapshot in trunk: 71d2dc5

fluiddot committed:
Update `@react-navigation/stack` to version `6.3.5`

fluiddot committed:
Disable `react-native-screens` in navigators

`react-native-screens` is meant to be used at root level to save memory when having inactive screens. This is not the case of the editor, as the stack navigators are used within the Bottom sheet component.
As a side note, enabling `react-native-screens` here leads to the editor crashing.

fluiddot committed:
Fix render order of animated view to highlight selected segment

Rendering the animated view before the segments will ensure that is rendered behind them.

fluiddot committed:
Upgrade `react-native-linear-gradient` to version `2.7.3`

This commit also updates the `react-native-hsv-color-picker` library to point to the same version of `react-native-linear-gradient`.

fluiddot committed:
Use `react-native-safe-area-context` mock provided by the library

fluiddot committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

# Conflicts:
#	package-lock.json

fluiddot committed:
Update link modal snapshot

fluiddot committed:
Update `package-lock.json` file

The integrity checksum of `react-native-hsv-color-picker` changed because the package has been modified (ref: https://github.com/wordpress-mobile/react-native-hsv-color-picker/pull/10#discussion_r1245039081)

fluiddot commented on test/native/setup.js:
```diff
@@ -168,21 +169,7 @@ jest.mock( 'react-native-safe-area', () => {
-		useSafeAreaFrame: jest.fn().mockImplementation( () => frame ),
-	};
-} );
+jest.mock( 'react-native-safe-area-context', () => mockSafeAreaContext );
```
Now that we use a newer version of `react-native-safe-area-context`, we can use the mock provided by the package ([reference](https://github.com/th3rdwave/react-native-safe-area-context#testing)).

fluiddot committed:
Update source of `react-native-hsv-color-picker` to use tag version

fluiddot committed:
Revert "Update link modal snapshot"

This reverts commit 7988b0ec41facd4dcbd95178b89afd8ffa9c4dc2.

This is needed after disabling `react-native-screens` in navigators (ref: https://github.com/WordPress/gutenberg/pull/51303/commits/e5838f473a8bb20c1df45db4a64c32ccbd7a693b).

fluiddot committed:
[RNMobile] Upgrade React Native `0.71.11` - iOS changes (#51386)

* refactor: Extract bundle version number to var

* refactor: Delete /.ruby-version, no longer needed

* refactor: Update /podfile to align w/RN updates

* refactor: Remove path names as part of RN upgrade

* Update `Podfile` with changes from RN upgrade helper

* Fix React Native path for `react_native_post_install` script

* Update Pods

* Add patch to fix Reanimated podspec

Without this patch, Reanimated tries to use Hermes version of React Native and produced a build failure. Seems there's an issue in the `podspec` file, as the JSC module is not being added.
Reference: https://github.com/software-mansion/react-native-reanimated/issues/4254

* Update pods to reflect 0.71.11 target

* Apply changes to pods following `pod install`

* Update `Podfile.lock` file

---------

Co-authored-by: Siobhan <siobhan@automattic.com>

fluiddot committed:
[RNMobile] Upgrade React Native `0.71.11` - Android changes (#51289)

* Upgrade Gradle to version 7.5.1

* Upgrade Gradle plugin

* Remove no longer needed files in new version

* Update Flipper initialization

* Update demo project main application

* Remove gradle download task plugin

* Bump ndk version

* Remove no longer need logic related to `newArchEnabled`

* Apply plugin React Native Gradle plugin

* Use React Native and Hermes modules from Maven

We no longer need to publish these binaries because React Native team is publishing React Native binaries to Maven.

* Remove exclude group from Flipper

* Update comments in `build.gradle` to align with new RN version

* Remove deprecated Gradle property

* Add `mavenLocal` repository to allow testing local binaries

* Bump Reanimated and Gesture handler libraries

* Revert "Upgrade Gradle plugin"

This reverts commit 82764a265d6b8ea57ea6b6c420b3e244d347ec03.

* build: Resolve react-native-gradle-plugin incompatability

Due to host app requirements, we must use AGP 7.2.1. The included patch
disables logic requiring AGP 7.3. The logic appears to not be required
for our use cases in the Demo editor or host apps.

We should remove this patch once we upgrade past AGP 7.3.

* Reduce priority of `mavenLocal` repository in Android build configurations

Maven local is used to provide dependencies located locally, which is mainly used for testing and debugging. Hence, published dependencies should be prioritized over local ones.

* Disable `react-native-screens` in navigators

`react-native-screens` is meant to be used at root level to save memory when having inactive screens. This is not the case of the editor, as the stack navigators are used within the Bottom sheet component.
As a side note, enabling `react-native-screens` here leads to the editor crashing.

* Fix render order of animated view to highlight selected segment

Rendering the animated view before the segments will ensure that is rendered behind them.

* Bump Linear gradient Android library

This library is now published via the `react-native-libraries-publisher` repository.

---------

Co-authored-by: David Calhoun <github@davidcalhoun.me>

fluiddot commented:
Now that all native changes have been incorporated we can set the PR ready for review. However, **note that we still have issues related to iOS in the Gutenberg Mobile PRs, so shouldn't merge this PR until they are ready.**

fluiddot committed:
Avoid exception in E2E tests when typing an empty string on Android

fluiddot committed:
Update a11y id queries for Android E2E tests

Starting in React Native 0.71, the accessibility hint is no longer appended to the accessibility content description. Reference: https://github.com/facebook/react-native/commit/0b70b38547685b6b978cacedf3ff36723d79a594

fluiddot committed:
Update button inline appender query for Android E2E tests

fluiddot committed:
Unify press keycode function for E2E tests

fluiddot commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -195,6 +188,9 @@ const stopDriver = async ( driver ) => {
  * On iOS: "clear" is not defaulted to true because calling element.clear when a text is present takes a very long time (approx. 23 seconds)
  */
 const typeString = async ( driver, element, str, clear ) => {
+	if ( isKeycode( str ) ) {
+		return await pressKeycode( driver, getKeycode( str ) );
+	}
```
When typing special characters like `\n`, instead of typing them into the text input, we'll execute them by pressing their associated keycode. For example, typing `\n` will result in pressing the Enter key. On Android, this behavior was already in place ([reference](https://github.com/WordPress/gutenberg/blob/1be8fa3c5f63a0c9a096478805665724417512aa/packages/react-native-editor/__device-tests__/helpers/utils.js#L233-L240)), but not on iOS as typing the new line character is breaking the line. However, after the RN upgrade, I noticed that on iOS now typing on an element always places the caret at the start. This is unexpected compared to the past behavior, where invoking `typeString` always appended the string to wherever the caret was previously placed.

I haven't explored further what changed after the RN upgrade, but I found it more reliable to use keycodes for special characters instead of typing them.

**Example:**

| After | Before |
|--------|--------|
| <video src=https://github.com/WordPress/gutenberg/assets/14905380/eacf3282-c12a-4cd1-bb3f-4d77a78bf380> | <video src=https://github.com/WordPress/gutenberg/assets/14905380/e7f3ea57-d939-4f4b-b6c8-0163ad2e553f> |

fluiddot committed:
Update comments in functions related to pressing a keycode

fluiddot committed:
Update block drop position using Reanimated's shared value

Seems there's some kind of incompatibility on calling a JS function from a worklet invoked from a gesture handler. For this reason, the logic to set the dropping insertion point has been updated. It now uses a Reanimated's shared value to keep the dragging over position and  `useDerivedValue` hook to listen for changes.

dcalhoun commented on packages/react-native-editor/ios/.ruby-version:
Do we believe this is safe to remove given [its origin](https://github.com/WordPress/gutenberg/commit/d52b240cbb0e5c5e1bf6789629b30121daac1615)?

fluiddot replied:
Ah, good observation. I removed this file following the RN upgrade helper, but as you mentioned, seems we might need it to match the Ruby version of the hosted app 👍. I'll revert the change and add an inline comment to avoid this in future RN upgrades.

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/cf67f4965ef86d829a480a80250f002bd1ce4d6e.

dcalhoun commented on patches/react-native-reanimated+2.17.0.patch:
```diff
@@ -0,0 +1,12 @@
+diff --git a/node_modules/react-native-reanimated/RNReanimated.podspec b/node_modules/react-native-reanimated/RNReanimated.podspec
```
WDYT about adding a note ([example](https://github.com/WordPress/gutenberg/blob/bb3afd8127ba7e8b1a7505217f26fb4e7b210f20/patches/react-native-gradle-plugin%2B0.71.19.patch#L1)) to the top of this file indicating when it might be removed?

fluiddot replied:
Good idea, I'll add a note.

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/a5036ddd783544a82a0ee66123a011ca59475786.

dcalhoun commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -238,35 +252,34 @@ describe( 'Paragraph block', () => {
+		fireEvent.press( screen.getByLabelText( 'Link' ) );
+
 		fireEvent.changeText(
 			screen.getByPlaceholderText( 'Add link text', { hidden: true } ),
```
The `hidden: true` option is not necessary now that we refactored this test to reorder the interaction steps. This relates to https://github.com/WordPress/gutenberg/pull/46735#discussion_r1232489273. 
```suggestion
			screen.getByPlaceholderText( 'Add link text' ),
```

dcalhoun commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -238,35 +252,34 @@ describe( 'Paragraph block', () => {
+		const typeURLInput = await waitFor( () =>
+			screen.getByPlaceholderText( 'Search or type URL' )
+		);
+		// Wait for side effects produced by link suggestions
+		await act( () =>
+			fireEvent.changeText( typeURLInput, 'wordpress.org' )
+		);
```
Similar to https://github.com/WordPress/gutenberg/commit/70dbcc60a1fdf4eb1f7ccb9eedf05e037b141bc0, we can avoid this generic flush of micro tasks by awaiting the removal of the loading indicator. 

```js
await waitForElementToBeRemoved( () =>
  screen.getByTestId( 'link-picker-loading' )
);
```

fluiddot replied:
Ah, interesting. I'll check this and the below comments related to integration tests 👍.

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/70c6dcb9e38fe3649079fd660e04c12bfdd84669.

dcalhoun commented on packages/block-library/src/paragraph/test/edit.native.js:
```diff
@@ -287,31 +300,29 @@ describe( 'Paragraph block', () => {
+		// Wait for side effects produced by link suggestions
+		await act( () =>
+			fireEvent.changeText( typeURLInput, 'wordpress.org' )
+		);
```
Similar to https://github.com/WordPress/gutenberg/commit/70dbcc60a1fdf4eb1f7ccb9eedf05e037b141bc0, we can avoid this generic flush of micro tasks by awaiting the removal of the loading indicator. 

```js
await waitForElementToBeRemoved( () =>
  screen.getByTestId( 'link-picker-loading' )
);
```

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/70c6dcb9e38fe3649079fd660e04c12bfdd84669.

dcalhoun commented on packages/components/src/mobile/link-settings/test/edit.native.js:
```diff
@@ -112,21 +128,25 @@ describe.each( [
-								: 'Search or type URL'
-						}`
+				fireEvent.press( subject.getByLabelText( 'Open Settings' ) );
+				// Wait for side effects produced by Clipboard
+				await act( () =>
+					fireEvent.press(
+						subject.getByLabelText(
+							`Link to, ${
+								type === 'core/image'
+									? 'None'
+									: 'Search or type URL'
+							}`
+						)
 					)
 				);
```
This wrapping `act` does not appear to be necessary and can be removed. I find this surprising as I opened https://github.com/WordPress/gutenberg/pull/51613 to specifically [address](https://github.com/WordPress/gutenberg/pull/51303#discussion_r1233062582) this `act` warning. Do you encounter `act` warnings if you remove this?

fluiddot replied:
Ah, good observation. I can't recall if I encountered `act` warnings, but it's likely that was the case. However, I've just tested removing it and tests are passing 🤔. Not sure if when I add it I was trying to solve an issue that got addressed differently. In any case, I'll proceed to remove the `act` blocks.

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/2d0a21954fe9417059b7566cb7146fd7d37e0835.

dcalhoun commented on packages/components/src/mobile/link-settings/test/edit.native.js:
```diff
@@ -112,21 +128,25 @@ describe.each( [
 				if ( type === 'core/image' ) {
-					fireEvent.press(
-						await subject.findByLabelText( /Custom URL/ )
+					// Wait for side effects produced by Clipboard and link suggestions
+					await act( () =>
+						fireEvent.press(
+							subject.getByLabelText( /Custom URL/ )
+						)
 					);
```
This wrapping `act` does not appear to be necessary and can be removed. I find this surprising as I opened https://github.com/WordPress/gutenberg/pull/51613 to specifically [address](https://github.com/WordPress/gutenberg/pull/51303#discussion_r1233062582) this `act` warning. Do you encounter `act` warnings if you remove this?

fluiddot replied:
Same as I shared in [this comment](https://github.com/WordPress/gutenberg/pull/51303#discussion_r1252904157). I'll proceed to remove it. Thanks!

fluiddot replied:
Suggestion applied in https://github.com/WordPress/gutenberg/pull/51303/commits/2d0a21954fe9417059b7566cb7146fd7d37e0835.

dcalhoun reviewed (commented):
It seems we are close. I did not provide an approval given we intend to await the other downstream branches as well. Let me know if we'd like to go ahead and approve this instead.

fluiddot commented:
Related to the crash mentioned [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5924#issuecomment-1619800846), I investigated it and found out that commenting [this line](https://github.com/WordPress/gutenberg/blob/b1579dd8d3bc745a1f25da2872f2df3cc16afc28/packages/block-editor/src/components/block-draggable/index.native.js#L187) prevented it. I tried different options to avoid it but seems there's some kind of incompatibility when calling a JS function from a worklet invoked in a gesture handler:

https://github.com/WordPress/gutenberg/blob/b1579dd8d3bc745a1f25da2872f2df3cc16afc28/packages/components/src/draggable/index.native.js#L109-L127

https://github.com/WordPress/gutenberg/blob/b1579dd8d3bc745a1f25da2872f2df3cc16afc28/packages/block-editor/src/components/block-draggable/index.native.js#L262-L264

https://github.com/WordPress/gutenberg/blob/b1579dd8d3bc745a1f25da2872f2df3cc16afc28/packages/block-editor/src/components/block-draggable/index.native.js#L180-L191

**Video capture:**

| Crash - Calling `onBlockDragOver` in JS thread | No crash - Commenting call to `onBlockDragOver` |
|--------|--------|
| <video src=https://github.com/WordPress/gutenberg/assets/14905380/7c161c76-0e0e-4643-af22-2caf23985c27> | <video src=https://github.com/WordPress/gutenberg/assets/14905380/2b0f4d00-23ee-4154-a274-992c0157ece2> | 

<hr>

In order to fix the crash, I opted for refactoring how the dropping insertion point is calculated (i.e. the logic of `onBlockDragOver` handler). I created a worklet to avoid calling the function in the JS thread. However, we still need to run code in the JS thread because the functions used in the calculations of the dropping insertion point need to happen on that thread. For this, what I did was listening for changes using [the `useDerivedValue` hook](https://docs.swmansion.com/react-native-reanimated/docs/2.x/api/hooks/useDerivedValue/) in a Reanimated's shared value, and invoke the function to do the calculations. As far as I tested, this approach seems to work as I couldn't reproduce the crash.

* Commit ref: https://github.com/WordPress/gutenberg/pull/51303/commits/bdd47040043bc862b313384cb556213b3b6dfbd1

As I shared [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5924#issuecomment-1619800846), seems that with the RN upgrade or enabling Hermes, we might have uncovered the culprit of crashes related to Reanimated in production versions. Hence, I'm planning to incorporate these changes in `trunk` via a separate PR and hopefully release it in the next version 🤞 .

fluiddot committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

# Conflicts:
#	packages/react-native-editor/ios/Podfile.lock

fluiddot committed:
Remove unneeded `hidden` param in Paragraph block test case

Co-authored-by: David Calhoun <github@davidcalhoun.me>

fluiddot committed:
Revert removing `.ruby-version` file

fluiddot committed:
Add inline comment in Reanimated patch

fluiddot committed:
Use `waitForElementToBeRemoved` in Paragraph block test cases

This way we can avoid waiting for any microtasks of link suggestions.

fluiddot committed:
Remove `act` statements from Link Settings test cases

dcalhoun committed:
fix: Cover focal point drag handle visibility

The lack of an explicit width or height resulted in a invisible drag
handle. The logic passing the dimensions to the SVG expected a single
style object. The reality is that it (1) referenced only the Sass styles
and (2) the combined reference was actually an array of style objects.

Updating the reference and flattening it ensures the appropriate width
and height are passed to the SVG.

It appears the absence of explicit dimensions was not an issue in
earlier versions of React Native, but it makes sense why it might be
required.

fluiddot committed:
[RNMobile] Use Reanimated in bottom sheet height animation (#52563)

* Expose max height properties in `BottomSheetProvider`

* Animate bottom sheet's height with Reanimated

Pass `currentHeight` in bottom sheet navigation context

* Use pixel value when setting fullscreen height

We need to pass pixel values in order to animate the height with Reanimated.

* Rename `heightRef` to `maxHeight`

* Re-enable `exhaustive-deps` lint rule in `BottomSheetNavigationContainer`

* Avoid setting height using debounce

* Add test ID to navigation container component

* Mock Reanimated's `now` function

* Update test cases related to bottom sheet height animation

* Update test snapshots

fluiddot committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

# Conflicts:
#	package-lock.json
#	packages/react-native-editor/__device-tests__/pages/editor-page.js
#	packages/react-native-editor/android/app/build.gradle
#	packages/react-native-editor/ios/Podfile.lock

fluiddot committed:
Update `react-native-editor` changelog

fluiddot committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

# Conflicts:
#	packages/block-library/src/audio/test/__snapshots__/edit.native.js.snap
#	packages/block-library/src/file/test/__snapshots__/edit.native.js.snap

SiobhyB reviewed (approved):
Verified I'm able to go through the testing steps by opening the editor and performing some straightforward smoke tests. 🙌 🎉

SiobhyB committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

fluiddot committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

SiobhyB committed:
Drop unsupported `--no-jetifier` from Android cmd

The `--no-jetifier` option no longer appears to be supported and results in an error when attempting to build the Android demo app.

Ref: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5881#issuecomment-1620435089

SiobhyB committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

SiobhyB committed:
Revert accidental change to .ruby-version

SiobhyB committed:
Restore correct dependencies to package-lock.json

fluiddot committed:
Update `react-native-editor` changelog

fluiddot committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

fluiddot commented:
Seems that a couple of the End-to-End Tests / Playwright CI jobs are failing consistently (I retried them three times). I don't think these changes are breaking the web version, so I'm going to update with `trunk` in case it's a known issue that has been already solved 🤞 .

fluiddot commented:
> Seems that a couple of the End-to-End Tests / Playwright CI jobs are failing consistently (I retried them three times). I don't think these changes are breaking the web version, so I'm going to update with `trunk` in case it's a known issue that has been already solved 🤞 .

Seems it keeps failing, I'll take a look and run them locally in case any of the changes of the PR introduced a regression.

fluiddot committed:
Merge branch 'trunk' into rnmobile/upgrade/react-native-0.71.8

fluiddot committed:
Update `package-lock.json` file to revert previous conflict resolutions

In https://github.com/WordPress/gutenberg/pull/51303/commits/4482b9df651d1aa6109efc1f0b5b2ddcb4116da0 we had a conflict in `package-lock.json` that was solved using the changes from this branch. However, seems that something went wrong and that although the editor has no issues, some e2e tests are failing due to this.

This has been solved by using the latest version of `package-lock.json` file from `trunk` and updating it with the package updates required in the React Native upgrade.

fluiddot committed:
Re-apply `react-devtools-core` patch to new version

fluiddot committed:
Update `Podfile.lock` file

fluiddot commented:
> > Seems that a couple of the End-to-End Tests / Playwright CI jobs are failing consistently (I retried them three times). I don't think these changes are breaking the web version, so I'm going to update with `trunk` in case it's a known issue that has been already solved 🤞 .
> 
> Seems it keeps failing, I'll take a look and run them locally in case any of the changes of the PR introduced a regression.

This issue should be addressed via https://github.com/WordPress/gutenberg/pull/51303/commits/c6b13112b80e5345425e5c8b99b090fe456d6dbd (further information can be found in the commit message). From this change, we had also updated a patch and the `Podfile.lock` file:
* https://github.com/WordPress/gutenberg/pull/51303/commits/29c2bedc2321bdc0ea96cb75fe293d55f2b9776b
* https://github.com/WordPress/gutenberg/pull/51303/commits/61669c81fb2b99f0fba188e2c3fe219b30f73718

priethor commented:
@fluiddot I see you experienced the new PR Label enforcer 😅 . I see it's frequent that mobile-app-related PRs don't have any other PRs. Should I add `Mobile App - i.e. Android or iOS` as the list of permitted types of PR so that they don't need any other label?

fluiddot commented:
> @fluiddot I see you experienced the new PR Label enforcer 😅 .

Yeah, it's a nice addition 🎊.

> I see it's frequent that mobile-app-related PRs don't have any other PRs.

Yes, it's not very common although it might be the case when working on a feature.

> Should I add `Mobile App - i.e. Android or iOS` as the list of permitted types of PR so that they don't need any other label?

@priethor To be honest, I think we can benefit from using the type label. This way we can add more context to the PRs 🙂. For now, I'd avoid adding the Mobile label. Thanks 🙇 !

priethor commented:
@fluiddot To clarify, you can (and probably should?) still add the Mobile label! The new pre-merge check ensures all PRs are of type enhancement, bugfix, etc. So nothing prevents adding a type like `Bugfix` together with `Mobile`. 😊

fluiddot commented:
> @fluiddot To clarify, you can (and probably should?) still add the Mobile label! The new pre-merge check ensures all PRs are of type enhancement, bugfix, etc. So nothing prevents adding a type like `Bugfix` together with `Mobile`. 😊

Oh sorry, I didn't want to imply that we won't use the Mobile label. I just wanted to mention that we can avoid adding that label to the list of permitted types 🙂. Thanks!

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into rnmobile/upgrade/react-native-0.71.8
